pysb/simulator/opencl_ssa.py

```diff
@@ -0,0 +1,357 @@
+
+    def _compile(self):
+
+        if self.verbose:
+            self._logger.info("Output OpenCl file to ssa_opencl_code.cl")
+            with open("ssa_opencl_code.cl", "w") as source_file:
+                source_file.write(self._code)
+
+        # allow users to select platform and devices
+        self.context = cl.create_some_context(True)
+        devices = self.context.devices
+        self._device_name = devices[0].name
+        self._n_devices = self.context.num_devices
+
+        # set local work group size. CPU = 1, CUDA = 32, otherwise 64
+        if devices[0].type == device_type.CPU:
+            self._local_work_size = (1, 1)
+        elif 'CUDA' in devices[0].platform.name.upper():
+            self._local_work_size = (32, 1)
+        else:
+            self._local_work_size = (64, 1)
+
+        self._logger.info(f"Using device {self._device_name}")
+        self.devices = devices
+
+        self.program = cl.Program(self.context, self._code).build(
+            options=[
+                # '-O3',
+                # '-cl-std=2.0',
+                # '-cl-uniform-work-group-size',
+                # '-cl-fast-relaxed-math',
+                # '-cl-single-precision-constant',
+                '-cl-denorms-are-zero',
+                '-cl-no-signed-zeros',
+                '-cl-finite-math-only',
+                '-cl-mad-enable',
+                '-I {}'.format(os.path.join(os.path.dirname(__file__)))
+            ]
+        )
+
+    def run(self, tspan=None, param_values=None, initials=None, number_sim=0):
+        """
+        Run a simulation and returns the result (trajectories)
+
+        .. note::
+            In early versions of the Simulator class, ``tspan``, ``initials``
+            and ``param_values`` supplied to this method persisted to future
+            :func:`run` calls. This is no longer the case.
+
+        Parameters
+        ----------
+        tspan
+        initials
+        param_values
+            See parameter definitions in :class:`ScipyOdeSimulator`.
+        number_sim: int
+            Number of simulations to perform
+
+        Returns
+        -------
+        A :class:`SimulationResult` object
+        """
+        super(OpenCLSSASimulator, self).run(tspan=tspan, initials=initials,
+                                            param_values=param_values,
+                                            number_sim=number_sim)
+        if tspan is None:
+            if self.tspan is None:
+                raise Exception("Please provide tspan")
+            else:
+                tspan = self.tspan
+        # tspan for each simulation
+        t_out = np.array(tspan, dtype=self._dtype)
+        # compile kernel and send parameters to GPU
+        if self._step_0:
+            self._setup()
+
+        self._logger.info("Creating content on device")
+        n_sim_per_device = self.num_sim//self._n_devices
+
+        if self.num_sim < self._local_work_size[0]:
+            local_work_size = (1, 1)
```

Comment:
why not use `self.num_sim` workers instead of just one?

Reply:
The rationale behind all these calculations around number of simulations, work size, blocks, threads, etc. is not obvious. A comment explaining _why_ (not what) would help a lot here. Then we can evaluate whether the code actually implements that design accurately.

Reply:
Added. Despite my best attempts, there is not an "exact" optimal number for work size, blocks, and number of threads. In an ideal world, when the GPUs are being used properly with no branching, there is. But in our case, there are ball park best, but the branching makes it trivial optimizations overall. I added the important parts, which basically try to make sure chunks of simulations go through as a warp size of 32 for NVIDIA, or wavefront size of 64 for AMD, but what really matters is that the work size or threads per block is a multiple of these. 
For small number of simulations (small than 32 or 64), it really doesn't matter what value you set, so I set to 1. But to simplify, I just reverted to be local_work_size. Since it generates up to a block, it pads the rest with 0s for either case.